### PR TITLE
replace v2 stripe js with v3

### PR DESCRIPTION
**Pull request by cmoylan**
https://www.pivotaltracker.com/story/show/175383754

Update Stripe JS to v3.

### CODE REVIEW

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 1]
```diff
@@ -1,66 +1,52 @@
...
-      return 'visa';
-    } else if (ccType === 'American Express') {
-      return 'amex';
-    } else if (ccType === 'Discover') {
-      return 'discover';
-    } else if (ccType === 'Diners Club') {
-      return 'dinersclub';
-    } else if (ccType === 'JCB') {
-      return 'jcb';
-    }
```
Comment (cmoylan) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb:26]: The new Stripe component handles this

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 2]
```diff
@@ -1,71 +1,57 @@
...
-      Spree.stripePaymentMethod.append("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][month]' value='" + response.card.exp_month + "'/>");
-      Spree.stripePaymentMethod.append("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][year]' value='" + response.card.exp_year + "'/>");
-      return Spree.stripePaymentMethod.parents("form").get(0).submit();
-    }
+    form.append($("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][gateway_payment_profile_id]' value='" + token.id + "'/>"));
+    form.append($("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][last_digits]' value='" + token.card.last4 + "'/>"));
+    form.append($("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][month]' value='" + token.card.exp_month + "'/>"));
+    form.append($("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][year]' value='" + token.card.exp_year + "'/>"));
```
Comment (jvanetten-printavo) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb]: I don't think `class` is needed?

A helper method would make this more readable and remove code duplication:
```
  function addInput(name, value) {
    form.append($("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][" + name + "]' value='" + value + "'/>"));
  }

  addInput("gateway_payment_profile_id",  token.id);
  addInput("last_digits",  token.card.last4);
  addInput("month",  token.card.exp_month);
  addInput("year",  token.card.exp_year);
```

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 3]
```diff
@@ -1,71 +1,57 @@
...
+        var errorElement = document.getElementById('card-errors');
+        errorElement.textContent = result.error.message;
+      } else {
+        // Send the token to your server
+        stripeTokenHandler(result.token);
       }
     });
+  };
+
+  $(document).ready(() => {
```
Comment (jvanetten-printavo) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb]: shorthand:

```suggestion
  $(() => {
```

**Review by jvanetten-printavo — CHANGES_REQUESTED** [review]

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 4]
```diff
@@ -1,71 +1,61 @@
...
-      Spree.stripePaymentMethod.append("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][last_digits]' value='" + response.card.last4 + "'/>");
-      Spree.stripePaymentMethod.append("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][month]' value='" + response.card.exp_month + "'/>");
-      Spree.stripePaymentMethod.append("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][year]' value='" + response.card.exp_year + "'/>");
-      return Spree.stripePaymentMethod.parents("form").get(0).submit();
-    }
+  function stripeTokenHandler(token) {
+    // Insert the token ID into the form so it gets submitted to the server
+    let form = $(Spree.stripePaymentMethod.parents("form").get(0));
+    let paymentMethodId = Spree.stripePaymentMethod.prop('id').split("_")[2];
+
```
Comment (jvanetten-printavo) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb:22]: js supports nested functions.  You could put the `addStripeHiddenInput()` here and you wouldn't need to pass `form` and `paymentMethodId` into it as those would be in scope.

Another option to not having to pass these in every time is to define `addStripeHiddenInput` as a closure  This code is only used in this method, so it makes sense to isolate it.   I think this makes more sense.

```
let form = $(Spree.stripePaymentMethod.parents("form").get(0));
let paymentMethodId = Spree.stripePaymentMethod.prop('id').split("_")[2];
let addStripeHiddenInput = (name, value) => form.append($("<input type='hidden' name='payment_source[" + paymentMethodId + "][" + name + "]' value='" + value + "'/>"));
addStripeHiddenInput("gateway_payment_profile_id",  token.id);
```

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 5]
```diff
@@ -1,71 +1,61 @@
...
+        var errorElement = document.getElementById('card-errors');
+        errorElement.textContent = result.error.message;
+      } else {
+        // Send the token to your server
+        stripeTokenHandler(result.token);
       }
     });
+  };
+
+  $(() => {
```
Comment (jvanetten-printavo) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb]: Move all the functions into this scope.  Right now, having functions at the `<script>` level makes them global on the window object.  In practice, we should avoid this as they can conflict with other libraries/code.

**Review by jvanetten-printavo — CHANGES_REQUESTED** [review]
Just some additional comments about where to put stuff.

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 6]
```diff
@@ -1,71 +1,59 @@
...
-      Spree.stripePaymentMethod.append("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][month]' value='" + response.card.exp_month + "'/>");
-      Spree.stripePaymentMethod.append("<input type='hidden' class='stripeToken' name='payment_source[" + paymentMethodId + "][year]' value='" + response.card.exp_year + "'/>");
-      return Spree.stripePaymentMethod.parents("form").get(0).submit();
-    }
-  };
+    function createToken() {
+      stripe.createToken(card).then((result) => {
+        if (result.error) {
+          // Inform the user if there was an error
+          var errorElement = document.getElementById('card-errors');
```
Comment (jvanetten-printavo) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb]:
```suggestion
          const errorElement = document.getElementById('card-errors');
```

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 7]
```diff
@@ -1,71 +1,59 @@
...
-</script>
-
+<script type="text/javascript" src="https://js.stripe.com/v3/"></script>
 <script>
-  Spree.stripePaymentMethod = $('#payment_method_' + <%= payment_method.id %>);
+  $(() => {
+    Spree.stripePaymentMethod = $('#payment_method_' + <%= payment_method.id %>);
+    let stripe = Stripe("<%= payment_method.preferred_publishable_key %>");
+    let elements = stripe.elements();
+    let card = elements.create('card', {
```
Comment (jvanetten-printavo) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb]:
```suggestion
    const stripe = Stripe("<%= payment_method.preferred_publishable_key %>");
    const elements = stripe.elements();
    const card = elements.create('card', {
```

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb [hunk 8]
```diff
@@ -1,71 +1,59 @@
...
+      }
+    });
+
+    function stripeTokenHandler(token) {
+      // Insert the token ID into the form so it gets submitted to the server
+      let form = $(Spree.stripePaymentMethod.parents("form").get(0));
+      let paymentMethodId = Spree.stripePaymentMethod.prop('id').split("_")[2];
+      let addStripeHiddenInput = (name, value) => {
+        form.append($("<input type='hidden' name='payment_source[" + paymentMethodId + "][" + name + "]' value='" + value + "'/>"));
+      }
```
Comment (jvanetten-printavo) [inline lib/views/frontend/spree/checkout/payment/_stripe.html.erb]: FYI, you don't need the `{}` for single-line closures.  Unless you had a reason?

```suggestion
      const form = $(Spree.stripePaymentMethod.parents("form").get(0));
      const paymentMethodId = Spree.stripePaymentMethod.prop('id').split("_")[2];
      const addStripeHiddenInput = (name, value) => form.append($("<input type='hidden' name='payment_source[" + paymentMethodId + "][" + name + "]' value='" + value + "'/>"));
```

**Review by jvanetten-printavo — CHANGES_REQUESTED** [review]
Sorry I missed the variable declarations could all be `const`

File: lib/views/frontend/spree/checkout/payment/_stripe.html.erb (hunk 8 above)
Comment (cmoylan) [reply lib/views/frontend/spree/checkout/payment/_stripe.html.erb]: I didn't like how long it was on the single line. That was the only reason

**Review by jvanetten-printavo — APPROVED** [review]

**Pull request merged**